pallets/wormhole/src/lib.rs

```diff
@@ -178,43 +116,106 @@ pub mod pallet {
+                &verifier.circuit_data.common,
+            )
+            .map_err(|e| {
+                log::error!("Proof deserialization failed. Error: {:?}", e);
+                Error::<T>::ProofDeserializationFailed
+            })?;
+
+            // Public inputs are ordered as follows:
+            // Nullifier.hash: 4 felts
+            // StorageProof.funding_amount: 2 felts
+            // StorageProof.root_hash: 4 felts
+            // ExitAccount.address: 4 felts
+            //
+            // TODO: These constants should be exposed from the common crate.
+            const PUBLIC_INPUTS_FELTS_LEN: usize = 14;
+            const NULLIFIER_START_INDEX: usize = 0;
+            const NULLIFIER_END_INDEX: usize = 4;
+            const FUNDING_AMOUNT_START_INDEX: usize = 4;
+            const FUNDING_AMOUNT_END_INDEX: usize = 6;
+            const EXIT_ACCOUNT_START_INDEX: usize = 10;
+            const EXIT_ACCOUNT_END_INDEX: usize = 14;
+
+            ensure!(
+                proof.public_inputs.len() == PUBLIC_INPUTS_FELTS_LEN,
+                Error::<T>::InvalidPublicInputs
+            );
+
+            let nullifier_bytes_vec =
+                felts_to_bytes(&proof.public_inputs[NULLIFIER_START_INDEX..NULLIFIER_END_INDEX]);
+            let nullifier_bytes: [u8; 32] = nullifier_bytes_vec
+                .try_into()
+                .map_err(|_| Error::<T>::InvalidPublicInputs)?;
 
             // Verify nullifier hasn't been used
             ensure!(
-                !UsedNullifiers::<T>::contains_key(public_inputs.nullifier),
+                !UsedNullifiers::<T>::contains_key(nullifier_bytes),
                 Error::<T>::NullifierAlreadyUsed
             );
 
-            VERIFIER_DATA.verify(proof).map_err(|_e| {
-                // log::error!("Verification failed: {:?}", e.to_string());
-                Error::<T>::VerificationFailed
-            })?;
+            verifier
+                .verify(proof.clone())
+                .map_err(|_| Error::<T>::VerificationFailed)?;
 
             // Mark nullifier as used
-            UsedNullifiers::<T>::insert(public_inputs.nullifier, true);
+            UsedNullifiers::<T>::insert(nullifier_bytes, true);
 
-            let exit_balance = public_inputs
-                .exit_amount
-                .try_into()
-                .map_err(|_| "Conversion from u64 to Balance failed")?;
+            let exit_balance_u128 = felts_to_u128(
+                <[F; 2]>::try_from(
+                    &proof.public_inputs[FUNDING_AMOUNT_START_INDEX..FUNDING_AMOUNT_END_INDEX],
+                )
+                .map_err(|_| Error::<T>::InvalidPublicInputs)?,
+            );
 
-            BalancesPallet::<T, ()>::mint_into(&public_inputs.exit_account, exit_balance)?;
+            // Check for overflow before converting to Balance type
+            let exit_balance: <T as BalancesConfig>::Balance =
+                if exit_balance_u128 > u64::MAX as u128 {
```

Comment:
we should just set Balances to be u128 in testing, this conditional shouldn't be here